src/waiter.c

```diff
@@ -0,0 +1,311 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+#include <stdlib.h>
+#include <stdalign.h>
+#include <stdbool.h>
+#include <inttypes.h>
+
+#include "macro_utils/macro_utils.h"
+
+#include "c_logging/xlogging.h"
+
+#include "c_pal/thandle.h"
+#include "c_pal/sm.h"
+#include "c_pal/srw_lock.h"
+
+#include "c_util/async_op.h"
+#include "c_util/rc_ptr.h"
+
+#include "c_util/waiter.h"
+
+typedef struct WAITER_TAG
+{
+    SM_HANDLE sm;
+    SRW_LOCK_HANDLE lock;
+    THANDLE(RC_PTR) data;
+    NOTIFICATION_CALLBACK notification_callback;
+    void* notification_callback_context;
+    NOTIFY_COMPLETE_CALLBACK notify_complete_callback;
+    void* notify_complete_callback_context;
+    bool registered;
+    bool notified;
+    bool callbacks_called;
+} WAITER;
+
+WAITER_HANDLE waiter_create(void)
+{
+    WAITER_HANDLE result;
+    WAITER_HANDLE waiter = malloc(sizeof(WAITER));
+    if (waiter == NULL)
+    {
+        LogError("Failure in malloc(sizeof(WAITER_HANDLE))");
+        result = NULL;
+    }
+    else
+    {
+        waiter->sm = sm_create("waiter");
+        if (waiter->sm == NULL)
+        {
```

Comment:
sm seems to be an overshoot, is it really needed?

Reply:
removed it